proto/coroutine/v1/coroutine.proto

```diff
@@ -45,6 +45,14 @@ message Region {
   // Type is the type of the region.
   uint32 type = 1;
 
+  // Array length, when >= 0, indicates that this region is an
+  // array of the specified length (in terms of number of
+  // elements, not in terms of bytes) and that the region type
+  // is the type of the array element. If the array length is < 0
+  // then this region does not represent an array, and the region
+  // type describes the object that is encoded.
+  int32 array_length = 2;
```

Comment:
Most types won't be array types, it's a bit unfortunate that we need to encode -1 on all other types (the zero value would be omitted from the wire representation). Would it be possible to use the `type` field to encode whether we are dealing with an array type? (e.g., use the least significant bit to say whether it's an array)

Reply:
FYI this is a `Region` not a `Type`. Putting the array length here on the region allows us to avoid creating a new type for the array that it represents. For example, all regions that are backing arrays for byte slices and strings point to the `uint8` type; we don't have to create and refer to a `[N]uint8` type for each unique N.

From an internal use case, there appears to be more regions that are backing arrays for strings and slices, compared to other types of memory regions such as structs and maps. We don't create a region for zero length arrays, so what we end up seeing is regions where most have `array_length > 0`. We could come up with a custom encoding for representing regions that aren't arrays, e.g. store this in the LSB of the region type ID, however we'd only save two bytes (PB tag + zig-zag-encoded -1) per non-array region which doesn't amount to much.

Reply:
Although, it does have extra savings because we can encode array length as an unsigned integer and avoid the zig-zag encoding.

Reply:
🤔 

I'm probably not familiar enough with the problem because I don't understand why we are concerned about encoding array `[N]byte`, for any sane program, all unique `N` should be known at compile time or bounded.

My understanding is we had created a problem for ourselves with the encoding of strings and slices as unique types, but generally it should be OK to represent array types of fixed size in the encoding if we can encode strings and slices as separate types?

Reply:
See b5a998a59b59155e90c9942cce8aaacc675b7c05.

Reply:
See f21a835

Reply:
OK, there are a few blind spots for me on this one but I can live with it 👍

Reply:
> I don't understand why we are concerned about encoding array [N]byte, for any sane program, all unique N should be known at compile time or bounded.

If you're dealing with untrusted input, and there are strings/bytes involved (either because you accept strings/bytes as input, or because you're doing work like making outbound requests to unknown locations and getting back strings/bytes), then `N` is not known at compile time. Hopefully `N` is at least bounded (e.g. you're rejecting inputs above a certain size, and putting limits on the size of headers/bodies/etc when making outbound requests), however you can still end up in a situation where `N` is large (e.g. you might reject requests once headers+body exceed megabytes) and you might end up with every `[X]uint8` type for `X=0..N`.

> My understanding is we had created a problem for ourselves with the encoding of strings and slices as unique types, but generally it should be OK to represent array types of fixed size in the encoding if we can encode strings and slices as separate types?

The reason we don't encode higher-level types like strings/slices as regions is because their backing arrays might be shared. We scan strings/slices to find those underlying memory regions, and we encode them as byte array regions. I think the Go language spec says that strings are immutable and cannot alias other strings/bytes, although it's possible to do this with `unsafe`. We could theoretically encode strings as their own regions, but that still wouldn't help us when encoding `[]byte` slices which may alias the backing arrays of other slices.

The serialized state contains types and data (in regions). This PR lets us avoid creating unique types for each unique byte array type (and `[N]uint8` types are unique per `N`). To do this, we store a bit more information when storing the data/regions, namely a length. It ends up looking a lot like how you would encode a higher-level object like a string (type=string + length + data).

Reply:
Thanks, makes a lot more sense now!